apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/source/GrpcRdaDLQSource.java

```diff
@@ -0,0 +1,360 @@
+package gov.cms.bfd.pipeline.rda.grpc.source;
+
+import static gov.cms.bfd.pipeline.rda.grpc.ProcessingException.isInterrupted;
+
+import com.codahale.metrics.Gauge;
+import com.codahale.metrics.Meter;
+import com.codahale.metrics.MetricRegistry;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Preconditions;
+import gov.cms.bfd.model.rda.MessageError;
+import gov.cms.bfd.pipeline.rda.grpc.NumericGauges;
+import gov.cms.bfd.pipeline.rda.grpc.ProcessingException;
+import gov.cms.bfd.pipeline.rda.grpc.RdaSink;
+import gov.cms.bfd.pipeline.rda.grpc.RdaSource;
+import io.grpc.CallOptions;
+import io.grpc.ManagedChannel;
+import java.time.Duration;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.function.BiPredicate;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+import javax.persistence.EntityManager;
+import lombok.Getter;
+import lombok.RequiredArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+
+/**
+ * DLQ specific RdaSource implementation that checks for failed messages in the {@link MessageError}
+ * table and re-queries them, then delegates actual service call and result mapping to other
+ * objects. This class has no dependency on a particular RPC or data type. It does maintain the
+ * ManagedChannel used to communicate with the gRPC service. The constructor creates the channel and
+ * the close() method closes the channel.
+ *
+ * <p>This object makes the RPC call and processes the stream. Stream processing consists of
+ * batching received objects and passing them to the RdaSink object for storage. Basic metrics are
+ * tracked at this level.
+ *
+ * @param <TMessage> type of objects returned by the gRPC service
+ */
+@Slf4j
+public class GrpcRdaDLQSource<TMessage, TClaim> implements RdaSource<TMessage, TClaim> {
```

Comment:
Maybe some of the common functionality from `GrpcRdaSource` could be refactored into an abstract base class?